pkg/controller/remotecluster/apikey.go

```diff
@@ -0,0 +1,190 @@
+		"local_namespace", reconciledES.Namespace,
+		"local_name", reconciledES.Name,
+		"remote_namespace", clientES.Namespace,
+		"remote_name", clientES.Name,
+	)
+
+	// Maintain a list of the expected API keys for that specific client cluster, to detect the ones which are no longer expected in the reconciled cluster.
+	expectedKeysInReconciledES := sets.New[string]()
+	// Same for the aliases
+	expectedAliases := sets.New[string]()
+	activeAPIKeysNames := activeAPIKeys.KeyNames()
+	for _, remoteCluster := range remoteClusters {
+		apiKeyName := fmt.Sprintf("eck-%s-%s-%s", clientES.Namespace, clientES.Name, remoteCluster.Name)
+		expectedKeysInReconciledES.Insert(apiKeyName)
+		expectedAliases.Insert(remoteCluster.Name)
+		if remoteCluster.APIKey == nil {
+			if activeAPIKeysNames.Has(apiKeyName) {
+				// We found an API key for that client cluster while it is not expected to have one.
+				// It may happen when the user switched back from API keys to the legacy remote cluster.
+				log.Info("Invalidating API key as remote cluster is not configured to use it", "alias", remoteCluster.Name)
+				if err := esClient.InvalidateCrossClusterAPIKey(ctx, apiKeyName); err != nil {
+					return err
+				}
+			}
+			continue
+		}
+
+		// Attempt to get an existing API Key with that key name.
+		activeAPIKey := activeAPIKeys.GetActiveKeyWithName(apiKeyName)
+		expectedHash := hash.HashObject(remoteCluster.APIKey)
+		if activeAPIKey == nil {
+			if err := createAPIKey(ctx, log, remoteCluster, apiKeyName, esClient, clientES, expectedHash, clientClusterAPIKeyStore, reconciledES); err != nil {
+				return err
+			}
+		} else {
+			// If an API key already exists ensure that the access field is the expected one using the hash
+			if err := maybeUpdateAPIKey(ctx, log, esClient, clientClusterAPIKeyStore, remoteCluster, activeAPIKey, apiKeyName, clientES, expectedHash); err != nil {
+				return err
+			}
+		}
+	}
+
+	// Get all the active API keys which have been created for that client cluster.
+	activeAPIKeysForClientCluster, err := activeAPIKeys.ForCluster(clientES.Namespace, clientES.Name)
+	if err != nil {
+		return err
+	}
+	// Invalidate all the keys related to that local cluster which are not expected.
+	for keyName := range activeAPIKeysForClientCluster.KeyNames() {
+		if !expectedKeysInReconciledES.Has(keyName) {
+			// Unexpected key, let's invalidate it.
+			log.Info("Invalidating unexpected API key", "key", keyName)
+			if err := esClient.InvalidateCrossClusterAPIKey(ctx, keyName); err != nil {
+				return err
+			}
+		}
+	}
+
+	// Delete all the keys in the keystore which are not expected.
+	aliases := clientClusterAPIKeyStore.ForCluster(reconciledES.Namespace, reconciledES.Name)
+	for existingAlias := range aliases {
+		if expectedAliases.Has(existingAlias) {
+			continue
+		}
+		clientClusterAPIKeyStore.Delete(existingAlias)
+	}
+
+	// Save the generated keys in the keystore.
+	if err := clientClusterAPIKeyStore.Save(ctx, c, clientES); err != nil {
```

Comment:
If I understand correctly we reconcile from the server's perspective which means if a cluster is client to multiple remotes multiple threads of reconciliation will try to update the same secret? Potentially at the same time depending on the parallelism the controller is configured with? Do you think this will be problem in practice? Did you consider doing it the other way round?

Reply:
> Did you consider doing it the other way round?

Yes, since we need an Elasticsearch client to reconcile the API keys it seemed more "natural" to me to create the ES client "once", and then reconcile/propagates the API keys for all the client clusters in the same loop, when the remote cluster is reconciled (as opposed to create _n_ Elasticsearch clients for each of the remote clusters to reconcile the API keys). My feeling was that, in case of a conflict on the `Secret`, the consequence would be that we would "just" need to create a new API key during the next iteration, and try to store it again, which seemed acceptable.

In any case one thing we may want to solve is when the API key is not immediately observed in the Secret (because of the client's cache). This would require a kind of "expectation mechanism": the generated keys would be stored in memory until they are persisted. This would prevent the conflict side effect when the `Secret` is created/updated since we would no longer need to generate a new key because we still have it in memory.

Reply:
Makes sense. It is probably computationally more expensive to create many ES clients with certificate pools than the current approach which minimises the number of clients created at the cost of potential conflicts.

Reply:
One downside of the current approach that occurred to me during testing is that the reconciliation that optimises for creating fewer clients for the remote server clusters is more disruptive to the client clusters as every new remote cluster connection requires a full restart of the cluster to update the keystores. I don't think this a problem in practice, it I can only think of very rare scenarios where this might be observable (adding remote clusters within very short intervals but not at once)